src/lib/paginator/paginator.ts

```diff
@@ -170,13 +170,13 @@ export class MatPaginator implements OnInit, OnDestroy {
 
   /** Whether there is a next page. */
   hasNextPage(): boolean {
-    const numberOfPages = this.getNumberOfPages();
-    return this.pageIndex < numberOfPages && this.pageSize != 0;
+    const maxPageIndex = this.getNumberOfPages() - 1;
+    return this.pageIndex < maxPageIndex && this.pageSize != 0;
   }
 
   /** Calculate the number of pages */
   getNumberOfPages(): number {
-    return Math.ceil(this.length / this.pageSize) - 1;
+    return Math.ceil(this.length / this.pageSize);
```

Comment:
Is `pageSize == 0` a thing we have to worry about here? (I see you checked for it above).